Improvements

dbcesar opened the pull request:
Depends on: 
https://github.com/rock-perception/perception-orogen-apriltags/pull/11

dbcesar committed:
Set velocities to zero when no marker is visible

dbcesar committed:
Transform the input setpoints from body to camera

dbcesar committed:
Fix timestamp of velocities output

Not it considers the time of the last received
virtual sample.

dbcesar committed:
Change the component to periodic instead of port_driven

This change aims to keep the consistence with the other
controllers.

dbcesar committed:
Create a property expected inputs to select the controlled DOF

- Select the variables on the jacobian matrix
- It writes NaN on the uncontrolled DOF
- Set to zero the undesirable input DOF, otherwise
  the component
  will fail when building the cdMo matrix.

dbcesar committed:
Create a property to set output saturation values

dbcesar committed:
Add a property to rotate the input data

Visp expect corners in a predefined order, however
on the enviroment this corners may come in another
order. This property aims to get over it.

Conflicts:
	visp.orogen

dbcesar committed:
Fix error on the input transformation

Use now the body to camera transformation in order to convert
the setpoints from the body to the camera frame

dbcesar committed:
Change the input port type to be compatible with apriltags

dbcesar committed:
Create a property for with data provided by a scaled image

dbcesar committed:
Create two properties to handle with desired targets objects

Those are the target_identifier, a string that will select
within the target_list the desired target to servoing.

dbcesar committed:
Add unit tests

doudou commented on tasks/Task.cpp:
```diff
@@ -107,14 +111,13 @@ void Task::updateHook()
-    bool no_setpoint = (_cmd_in.read(setpoint) == RTT::NoData);
-    ctrl_state.timestamp = base::Time::now();
+    bool no_corners = (_marker_corners.read(corners_vector) != RTT::NewData);
+    bool no_setpoint = (_cmd_in.read(setpoint) != RTT::NewData);
```
Why do you require a new setpoint each time ?

Apart from breaking general conventions (unless in a nested control loop, setpoints are either sent only once, or at low frequency), it creates a race condition between the corners and setpoint: you could read the setpoint in one updateHook and the corner in another, leading to not doing anything.

dbcesar replied:
I think now it is better

doudou commented on tasks/Task.cpp:
```diff
@@ -129,28 +132,40 @@ void Task::updateHook()
 
     if (state() == CONTROLLING)
```
It's really hard to follow long branches. So, don't do

```
if (X)
{
  VERY LONG PIECE OF CODE
}
else
{
  VERY SMALL PIECE OF CODE
}
<END OF FUNCTION>
```

Instead do

```
if (! X) {
  SMALL PIECE OF CODE
  return;
}

LONG PIECE OF CODE
<END OF FUNCTION>
```

dbcesar replied:
Done

doudou commented on tasks/Task.cpp:
```diff
@@ -283,14 +300,37 @@ void Task::setGain()
+        // linear
+        if (!expected_inputs.linear[i])
+        {
+            v[i] = base::NaN<double>();
```
Is there really a reason why you first update the `v` array and only then write in `vel` ? You could directly write `vel` ... (which would arguably be more readable ...)

dbcesar replied:
No, there is not.  Changed.

doudou commented on tasks/Task.cpp:
```diff
@@ -301,13 +341,135 @@ base::samples::RigidBodyState Task::convertToRbs(vpHomogeneousMatrix pose)
+        {
+            if ( i == 0 || i == 1)
+            {
+                cmd_in_body.angular[i] = -M_PI_2;
```
You really would need to explain that one with a comment.

dbcesar replied:
Done

doudou replied:
**Formatting** If you have special cases for two cases out of three .... Either use a separate default_angular_values array that you use in the loop. Or don't use a loop at all. Ifs are really hindering understanding.

**However** .... This `-M_PI_2` according to your comment is basically applying the frame-to-body transformation. Which you apply further down. So I still don't get it.

dbcesar replied:
>  However .... This -M_PI_2 according to your comment is basically applying the frame-to-body transformation. Which you apply further down. So I still don't get it.

The -M_PI_2 indeed is related to the frame-to-body transformation, however the loop is just to assing real values to the ones that are NaN (since they are not expected inputs). This is done in order to allow the Homogeneus Matrix creation.

doudou replied:
Still don't get it.

The normal workflow should be

input_values -> normalized_input_values -> transformation -> homogeneous matrix.

I.e. replace the NaN by values that make sense in _the input value space_ and then apply the transformation.

Having two different code paths for the NaN and normal cases spells trouble. In general, the less code paths there is, the better.

dbcesar replied:
Not sure If I understood, because I think I'm following the suggested workflow:

> input_values -> normalized_input_values -> transformation -> homogeneous matrix.

_Input values_ : the argument "setpoint"-> 
_normalized_input_values_: the "_for_" loop -> 
_transformation_: building the homogeneus matrix from the setpoint (bdMo)and multiplying by the transformation matrix (cMb) -> 
_homogeneus matrix_: the result of the transformation.

Am I missing something?

> Having two different code paths for the NaN and normal cases spells trouble. In general, the less code paths there is, the better.

I don't see how to have only a single path in this case. Any suggestions?

doudou replied:
Let's assume that I'm the one missing something for now ...
- in what frame is the setpoint expressed ?

dbcesar replied:
The _setpoint_ is the object pose wrt body frame .

However,  visp library works with the object pose in the _camera_ frame and that is why I use the body-to-camera matrix to transform it.

doudou commented on tasks/Task.cpp:
```diff
@@ -301,13 +341,135 @@ base::samples::RigidBodyState Task::convertToRbs(vpHomogeneousMatrix pose)
+    //bdMo is the desired pose of the object in the body frame
+    //cdMo is the desired pose of the object in the camera frame
+    //cMb is the transformation matrix of the body in the camera frame
+    cdMo = cMb * bdMo;
```
There's really no good reason to store this as an attribute. The only place where you use it is just after you called transformInput. Make it the return value of the method and use it as a local variable.

dbcesar replied:
Acctualy, I also use it on the line 264.

```
//////////////////////////////
// Update Feature 3 - t, tu //
//////////////////////////////
cdMc = cdMo*cMo.inverse();
t.buildFrom(cdMc);
tu.buildFrom(cdMc) 
```

doudou replied:
This method is part of the same code flow. Return the matrix from both transformInput (which should really be called computeDesiredPose) and then from updateDesiredPose (hell, maybe there's something wrong with this naming scheme ... IMO you should just merge both methods together, the latter really just "finishes" the former's job). Then, pass it as argument.

Generally speaking, avoid storing as attribute anything that does not need to be long-lived. And avoid side-effects in methods as long as it does not complexify the code base too much.

A method that computes a single value (like this one) should really be side-effect free (i.e. const) and return its result. Let the caller store it somewhere if it really has to (in this case, it does not). Doing so improves readability by making the data flow obvious (it would be obvious in updateHook that you compute the new desired pose and then compute the command from the new desired pose).

Finally, drop the boolean return values. All of them (except the ones from the `*Hook` methods that are mandated by RTT, obviously). Use exceptions if there are errors. You need to learn the reflex of asking yourself what was the point of having `bool` return values everywhere if **you never test against them**

dbcesar replied:
I merged the methods and now its way better.

I also tried to eliminate unecessary attributes, but still not sure if I can get rid of someone else. 
I tried also to turn the methods as much side effect free as possible, let me now if I can improve something else.

Concerning the bool methods...I took it out and now throw exeptions when it goes to an undesired behaviour. However I think I should handle  it somehow. Would you recomnend the creation of one more runtime state (for instance :WAITING_VALID_TARGET) and move the component there anytime these exception are thrown?

doudou replied:
> WAITING_VALID_TARGET

You can wait for a target. You should complain loudly (i.e. raise an exception **and terminate**) if you get an **invalid** target.

doudou commented on tasks/Task.cpp:
```diff
@@ -301,13 +341,135 @@ base::samples::RigidBodyState Task::convertToRbs(vpHomogeneousMatrix pose)
+
+    for(int i = 0; i < 3; ++i)
+    {
+        if (!expected_inputs.linear[i])  {eJe[i][i] = 0;}
```
This is a really weird style.

If you want to have it a bit compressed, simply use the syntax without `{}`

```
if (bla)
   blo;
```

Avoid one-liner, it makes code coverage difficult to understand.

dbcesar replied:
Changed here. Is there a better choice between

```
if  (bla)
    blo;
```

or 

```
if (bla)
{
    blo;
}

```

?

doudou replied:
I personally prefer the first.

doudou commented on tasks/Task.cpp:
```diff
@@ -301,13 +337,144 @@ base::samples::RigidBodyState Task::convertToRbs(vpHomogeneousMatrix pose)
+    //in order to have valid values for the HomogeneusMatrix creation
+    for (int i = 0; i < 3; ++i)
+    {
+        if (!expected_inputs.linear[i]) {cmd_in_body.linear[i] = 0;}
```
Single line ifs should be formatted

```
if (X)
  do_y;
```

Single-line stuff is harmful in general, as it hinders code coverage measurement (there are two separate branches on the same line)

Generally speaking, be consistent with yourself and with the rest of the codebase. This is the only place where you format a single-line if like this. In all other cases you use

```
if (X)
{
   do_y
}
```

doudou commented on tasks/Task.cpp:
```diff
@@ -283,14 +299,34 @@ void Task::setGain()
+        {
+            vel.linear[i] = base::NaN<double>();
+        }
+        else if (v[i] > saturation.linear[i])
```
what about negative saturations ???

dbcesar replied:
Indeed. Fixed it.

dbcesar committed:
Allow to add negative saturation values

doudou commented on tasks/Task.cpp:
```diff
@@ -31,25 +30,31 @@ bool Task::configureHook()
+    // radial values are informed.
     frame_helper::CameraCalibration cal = _camera_parameters.get();
+    double scaling = _scaling.get();
+    if(scaling != 1.0)
```
Don't. There's no harm in multiplying by one, get rid of the if.

doudou commented on tasks/Task.cpp:
```diff
@@ -127,30 +129,46 @@ void Task::updateHook()
     }
 
-    if (state() == CONTROLLING)
+    if (!(state() == CONTROLLING))
```
This is `state() != X`

doudou commented on tasks/Task.cpp:
```diff
@@ -127,30 +129,46 @@ void Task::updateHook()
-        // qdot = -lambda * L^+ * cVe * eJe * (s-s*)
-        task.set_eJe(eJe) ;
+    // filter the corners of an desired target
+    Points corners;
```
Point corners = value; instead of `Point corners; corners = value;`

doudou commented on tasks/Task.cpp:
```diff
@@ -127,30 +129,46 @@ void Task::updateHook()
+    // transforms the setpoint from the body frame to the camera frame and
+    // build the Homogeneus Matrix of the desired object pose in the camera
+    // frame
+    vpHomogeneousMatrix cdMo;
```
Same. Initialize inline instead of doing `init; assignation;`

doudou commented on tasks/Task.cpp:
```diff
@@ -237,77 +258,225 @@ void Task::updateFeatures(std::vector<base::Vector2d> corners)
+    /**
+     *  Update Feature 3 - t, tu
+     */
     cdMc = cdMo*cMo.inverse();
```
Any point in having this one stored as attribute ?

dbcesar committed:
Perform a general refactoring

- Exclude some unnecessary attributes
- Add a brief comment text for each method
- Declare methods and arguments as const, in order to avoid side-effects
- Perform the input ports reading in separeted methods
- Filter the targets and update their parameters at once and do this
  processing before the CONTROLLING state changing

dbcesar committed:
Add the repository README.md file

dbcesar commented:
@doudou I've added the documentation explain the conventions of this component. Especially the cmd_in (setpoint) case. 

I've also changed the way of find and update the target parameters. I've tried to follow your suggestion of return the vector index, but handle the cases on the updateHook (on the states checking) became more complex. So I've created two methods to read the input ports. 
In the case of the corners vector reading, it returns a boolean if either there is no new data on the input port or the desired target is not detected.

dbcesar commented:
@doudou ping

dbcesar commented:
ping @doudou

doudou commented:
:+1: Sorry for the delay in answering

dbcesar commented:
Sorry, but I couldn't track your answering.

dbcesar commented:
Sorry, but I couldn't track your answering.

dbcesar commented:
Any new comments here?

doudou commented:
Nope. That was what the thumbs-up meant ;-)
